crates/libcontainer/src/container/container_checkpoint.rs

```diff
@@ -0,0 +1,124 @@
+
+use libcgroups::common::{
+    CgroupSetup::{Hybrid, Legacy},
+    DEFAULT_CGROUP_ROOT,
+};
+use oci_spec::runtime::Spec;
+use std::os::unix::io::AsRawFd;
+
+const CRIU_CHECKPOINT_LOG_FILE: &str = "dump.log";
+
+impl Container {
+    pub fn checkpoint(&mut self, opts: &CheckpointOptions) -> Result<()> {
+        self.refresh_status()
+            .context("failed to refresh container status")?;
+
+        // can_pause() checks if the container is running. That also works for
+        // checkpoitning. is_running() would make more sense here, but let's
+        // just reuse existing functions.
+        if !self.can_pause() {
+            bail!(
+                "{} could not be checkpointed because it was {:?}",
+                self.id(),
+                self.status()
+            );
+        }
+
+        let mut criu = rust_criu::Criu::new().unwrap();
+
+        // We need to tell CRIU that all bind mounts are external. CRIU will fail checkpointing
+        // if it does not know that these bind mounts are coming from the outside of the container.
+        // This information is needed during restore again. The external location of the bind
+        // mounts can change and CRIU will just mount whatever we tell it to mount based on
+        // information found in 'config.json'.
+        let source_spec_path = self.bundle().join("config.json");
+        let spec = Spec::load(&source_spec_path)?;
+        let mounts = spec.mounts().clone();
+        for m in mounts.unwrap() {
+            match m.typ().as_deref() {
+                Some("bind") => {
+                    let dest = m
+                        .destination()
+                        .clone()
+                        .into_os_string()
+                        .into_string()
+                        .unwrap();
+                    criu.set_external_mount(dest.clone(), dest);
+                }
+                Some("cgroup") => {
+                    match libcgroups::common::get_cgroup_setup()
+                        .context("failed to determine cgroup setup")?
+                    {
+                        // For v1 it is necessary to list all cgroup mounts as external mounts
+                        Legacy | Hybrid => {
+                            for mp in libcgroups::v1::util::list_subsystem_mount_points().unwrap() {
+                                let cgroup_mount =
+                                    mp.clone().into_os_string().into_string().unwrap();
+                                if cgroup_mount.starts_with(DEFAULT_CGROUP_ROOT) {
+                                    criu.set_external_mount(cgroup_mount.clone(), cgroup_mount);
+                                }
+                            }
+                        }
+                        _ => (),
+                    }
+                }
+                _ => (),
+            }
+        }
+
+        let directory = std::fs::File::open(&opts.image_path)?;
+        criu.set_images_dir_fd(directory.as_raw_fd());
```

Comment:
Do we need to create the directory specified with `image_path`?

Reply:
+1

Reply:
We do not need to create it, but we could if it does not exist. I think in Podman we create it to make sure it gets created with 0600. We could do it here. Or not. I have not strong opinion. If it does not exist the user will see an error. By creating the directory we could avoid that error. On the other hand, if it does not exist then the user has to make a conscious decision where the checkpoint directory is. I have a couple of checkpoint directories all around my development system because the directory is automatically created. Both options have advantages and disadvantages.

We could leave it for now and if users are complaining then it can still be added later. Or if one of the container engines needs it, it can also be added later.